common/persistence/nosql/nosqlExecutionStore.go

```diff
@@ -0,0 +1,831 @@
+// Copyright (c) 2017-2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package nosql
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	p "github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/persistence/nosql/nosqlplugin"
+	"github.com/uber/cadence/common/persistence/nosql/nosqlplugin/cassandra"
+	"github.com/uber/cadence/common/persistence/nosql/nosqlplugin/cassandra/gocql"
+	"github.com/uber/cadence/common/types"
+)
+
+type (
+	// Implements ExecutionStore
+	nosqlExecutionStore struct {
+		shardID int
+		nosqlStore
+	}
+)
+
+var _ p.ExecutionStore = (*nosqlExecutionStore)(nil)
+
+// Guidelines for creating new special UUID constants for all NoSQL
+// For DB specific constants(e.g. Cassandra), create them in the db specific package. 
+// Each UUID should be of the form: E0000000-R000-f000-f000-00000000000x
+// Where x is any hexadecimal value, E represents the entity type valid values are:
+// E = {DomainID = 1, WorkflowID = 2, RunID = 3}
+// R represents row type in executions table, valid values are:
+// R = {Shard = 1, Execution = 2, Transfer = 3, Timer = 4, Replication = 5, Replication_DLQ = 6, CrossCluster = 7}
+const (
+	// Special Domains related constants
+	emptyDomainID = "10000000-0000-f000-f000-000000000000"
+	// Special Run IDs
+	emptyRunID                   = "30000000-0000-f000-f000-000000000000"
+	rowTypeReplicationWorkflowID = "20000000-5000-f000-f000-000000000000"
+	rowTypeReplicationRunID      = "30000000-5000-f000-f000-000000000000"
+	emptyInitiatedID             = int64(-7)
+)
+
+// NewExecutionStore is used to create an instance of ExecutionStore implementation
+func NewExecutionStore(
+	shardID int,
+	client gocql.Client,
+	session gocql.Session,
+	logger log.Logger,
+) (p.ExecutionStore, error) {
+	db := cassandra.NewCassandraDBFromSession(client, session, logger)
```

Comment:
nit: shall we add some TODO here indicating we need to support other nosql plugin as well here?

Reply:
yeah good catch.